File: 2022/day_01.py
def part1(data: str):
    elves = data.split("\n\n")
    elves = list(map(sum, map(lambda x: map(int, x.split("\n")), elves)))
    return max(elves)


def part2(data: str):
    elves = data.split("\n\n")
    elves = list(map(sum, map(lambda x: map(int, x.split("\n")), elves)))
    return sum(sorted(elves, reverse=True)[:3])


if __name__ == "__main__":
    test = True
    test = False
    test_input = """1000
2000
3000

4000

5000
6000

7000
8000
9000

10000"""
    if test:
        puzzle_input = test_input
    else:
        with open("day_01_input.txt", "r") as input_file:
            puzzle_input = input_file.read().strip()
    print(part1(puzzle_input))
    print(part2(puzzle_input))

File: 2018/day_01.rs
//! https://adventofcode.com/2018/day/1
//! https://adventofcode.com/2018/day/1/input

use std::{fs::read_to_string, time::Instant};

fn parse(input: &str) -> Vec<isize> {
    input.lines().map(|line| line.parse().unwrap()).collect()
}

pub mod part1 {
    use crate::day_01::parse;

    pub fn solve(input: &str) -> isize {
        parse(input).iter().sum()
    }
}

pub mod part2 {
    use std::collections::HashSet;

    use crate::day_01::parse;

    pub fn solve(input: &str) -> isize {
        let frequencies = parse(input);
        let mut seen = HashSet::new();
        let mut frequency = 0;
        let mut i = 0;
        loop {
            frequency += frequencies[i];
            if seen.contains(&frequency) {
                return frequency;
            }
            seen.insert(frequency);
            i = (i + 1) % frequencies.len();
        }
    }
}

pub fn main(test: bool) {
    let test_input = "".to_owned();
    let puzzle_input = if test {
        test_input
    } else {
        read_to_string("inputs/day_01_input.txt").unwrap()
    };
    let start = Instant::now();
    println!("{}", part1::solve(&puzzle_input));
    println!("Run in {:?}", start.elapsed());
    let start = Instant::now();
    println!("{}", part2::solve(&puzzle_input));
    println!("Run in {:?}", start.elapsed());
}
